### ISSUE TITLE
[regression] [17.7] disabling the 'quick' keyword in floating rules does not work anymore

### ISSUE THREAD

**Issue by fraenki** [opnsense/core#1745]
After upgrading from 17.1.11 to 17.7 I've noticed that all firewall rules, where a Gateway is defined, no longer work as expected. The Gateway directive seems to be ignored. I'm still seing the `route-to` directive in rules.debug...

It should be noted that the Gateway in this case is connected through OpenVPN.

**Comment by fraenki** [comment]
Hm, interesting, the _real_ issue is something different. It seems that the order in which rules are processed has changed. Now some of my (old) floating rules match first, so that my gateway rules never match.

Is this something intentional?

**Comment by fichtner** [comment]
It depends on what the actual issue is. Floating rules should always match first, so that sounds reasonable.

One thing we've changed is not redirecting outgoing traffic from an interface itself through the respective gateway anymore, unless a specific gateway is set in the floating rules. This was done to allow better multi-wan capabilities for squid, which always seemed to have suffered from this limitation.

**Comment by fraenki** [comment]
In this example I was using the gateway option in firewall rules to make sure certain outgoing traffic passes a OpenVPN tunnel (gateway) instead of my default WAN gateway.

bq. One thing we've changed is not redirecting outgoing traffic from an interface itself through the respective gateway anymore, unless a specific gateway is set in the floating rules.

Hm, I'm not sure if I fully understand this change. Does this mean that the gateway setting in _normal_ firewall rules was is less functional now? Do I _need_ to use floating rules for this kind of stuff from now on?

Can you point me to the issue/commit that introduced this change? Maybe it'll help me to better understand the underlying issue.

For me this turned out to be a breaking change. I'm not entirely sure this side-effect was intended. Maybe add a clear note to the 17.7 release notes or errata regarding this change and it's side-effects?

**Comment by AdSchellevis** [comment]
@fraenki I think you're looking for this one https://github.com/opnsense/core/issues/1681, if you send me both versions of rules.debug (17.1.11 and 17.7) maybe I an easily check if this is related.

**Comment by fraenki** [comment]
Got it! Floating rules have the keyword `quick` added in 17.7, which was missing in previous releases. For example:

```
# floating rule on 17.1.x
pass  on {  vtnet0  vtnet1  vtnet2  } inet proto tcp  from $local_networks to any port $http_https flags S/SA keep state  label "USER_RULE: allow HTTP/HTTPS to any"

# floating rule on 17.7
pass   quick on vtnet0 inet proto tcp from $local_networks to {any}  port $http_https keep state label "USER_RULE: allow HTTP/HTTPS to any" 
pass   quick on vtnet1 inet proto tcp from $local_networks to {any}  port $http_https keep state label "USER_RULE: allow HTTP/HTTPS to any" 
pass   quick on vtnet2 inet proto tcp from $local_networks to {any}  port $http_https keep state label "USER_RULE: allow HTTP/HTTPS to any" 
```

I have double-checked this: On 17.1.x the `quick` keyword was missing on this rule (and other rules too) both in `rules.debug` as well as `config.xml`. I've even restored a backup to check if the GUI option for `quick` was disabled.

**Comment by fraenki** [comment]
> I think this is a bug that needs fixing

Confirmed. On 17.7 it's not possible to disable `quick` on floating rules.
I'm not too familiar with this code and was not able to quickly spot the bug in the code :(

**Comment by fichtner** [comment]
Thanks for the find! The timing is a bit unfortunate. I'll have to add this as errata to the release announcement (opnsense-patch) and fix this with 17.7.1 later this week.
